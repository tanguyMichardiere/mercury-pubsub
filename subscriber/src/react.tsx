import type { ReactNode } from "react";
import { createContext, useContext, useEffect, useMemo } from "react";

import type { Channels } from "@mercury-pubsub/types";
import useEvent from "react-use-event-hook";

import type { SubscribeOptions } from ".";
import Subscriber from ".";

const subscriberContext = createContext<Subscriber | undefined>(undefined);

export default function SubscriberProvider(props: {
  url: string;
  _key: string;
  children: ReactNode;
}): JSX.Element {
  const subscriber = useMemo(() => new Subscriber(props.url, props._key), [props.url, props._key]);

  return (
    <subscriberContext.Provider value={subscriber}>{props.children}</subscriberContext.Provider>
  );
}

export function useSubscribe<C extends keyof Channels>(
  channel: C,
  events: Omit<SubscribeOptions<C>, "signal">
): void {
  const subscriber = useContext(subscriberContext);

  if (subscriber === undefined) {
    throw new Error("no subscriber context");
  }

  const abortController = useMemo(() => new AbortController(), []);

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onopen = useEvent(events.onopen ?? async function () {});
  const ondata = useEvent(events.ondata);
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onclose = useEvent(events.onclose ?? function () {});

  useEffect(
    function () {
      subscriber
        .subscribe(channel, { signal: abortController.signal, onopen, ondata, onclose })
        .catch(function (error) {
          console.error(error);
        });
      return function () {
        abortController.abort();
      };
    },
    [subscriber, channel, abortController, onopen, ondata, onclose]
  );
}
