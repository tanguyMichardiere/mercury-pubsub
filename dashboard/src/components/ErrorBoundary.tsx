import type { ErrorInfo, ReactNode } from "react";
import { Component } from "react";

import Link from "next/link";

type Props = {
  children: ReactNode;
};

type State = {
  error: unknown;
};

export default class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { error: undefined };
  }

  static getDerivedStateFromError(error: unknown): State {
    return { error };
  }

  componentDidCatch(error: unknown, errorInfo: ErrorInfo): void {
    console.error({ error, errorInfo });
  }

  reload(this: void): void {
    location.reload();
  }

  render(): JSX.Element {
    if (this.state.error !== undefined) {
      return (
        <div className="flex h-full flex-col items-center justify-center">
          <h2>An expected error has occurred, it has been reported</h2>
          <Link href="/">Home Page</Link>
          <button onClick={this.reload} type="button">
            Reload
          </button>
        </div>
      );
    }

    return <>{this.props.children}</>;
  }
}
