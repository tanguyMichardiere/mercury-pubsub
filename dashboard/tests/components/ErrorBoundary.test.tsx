import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";

import ErrorBoundary from "../../src/components/ErrorBoundary";

console.error = jest.fn();

const reload = jest.fn();
const location = window.location;
// @ts-expect-error necessary for mocking window.location methods
delete window.location;
window.location = { ...location, reload };

function ErrorThrower(): JSX.Element {
  throw Error();
}

describe("ErrorBoundary", function () {
  it("renders unchanged", function () {
    const tree = renderer
      .create(
        <ErrorBoundary>
          <ErrorThrower />
        </ErrorBoundary>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("allows to reload the page", function () {
    render(
      <ErrorBoundary>
        <ErrorThrower />
      </ErrorBoundary>
    );
    fireEvent.click(screen.getByText(/reload/i));
    expect(reload).toHaveBeenCalledTimes(1);
  });
});
