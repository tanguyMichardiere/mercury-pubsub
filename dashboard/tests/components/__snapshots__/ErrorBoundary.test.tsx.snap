// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary renders unchanged 1`] = `
<div
  className="flex h-full flex-col items-center justify-center"
>
  <h2>
    An expected error has occurred, it has been reported
  </h2>
  <a
    href="/"
  >
    Home Page
  </a>
  <button
    onClick={[Function]}
    type="button"
  >
    Reload
  </button>
</div>
`;
